src/cpp/rtps/transport/TCPTransportInterface.cpp

```diff
@@ -693,87 +688,87 @@ bool TCPTransportInterface::OpenOutputChannel(
+                locator) << ") @ " << IPLocator::to_string(locator));
 
-        auto channel_resource = channel_resources_.find(physical_locator);
+    std::lock_guard<std::mutex> socketsLock(sockets_map_mutex_);
+    auto channel_resource = channel_resources_.find(physical_locator);
 
-        // Maybe as WAN?
-        if (channel_resource == channel_resources_.end() && IPLocator::hasWan(locator))
+    // Maybe as WAN?
+    if (channel_resource == channel_resources_.end() && IPLocator::hasWan(locator))
+    {
+        Locator wan_locator = IPLocator::WanToLanLocator(locator);
+        channel_resource = channel_resources_.find(IPLocator::toPhysicalLocator(wan_locator));
+        if (channel_resource != channel_resources_.end())
         {
-            Locator wan_locator;
-            wan_locator.kind = locator.kind;
-            wan_locator.port = locator.port; // Copy full port
-            IPLocator::setIPv4(wan_locator, IPLocator::toWanstring(locator)); // WAN to IP
-            channel_resource = channel_resources_.find(IPLocator::toPhysicalLocator(wan_locator));
-            if (channel_resource != channel_resources_.end())
-            {
-                channel_resources_[IPLocator::toPhysicalLocator(locator)] = channel_resource->second; // Add alias!
-            }
+            channel_resources_[physical_locator] = channel_resource->second;     // Add alias!
         }
+    }
 
+    // (Server-Client Topology OR LARGE DATA with PDP discovery after TCP connection) - Server side
+    if (channel_resource != channel_resources_.end())
+    {
         std::shared_ptr<TCPChannelResource> channel;
-
-        if (channel_resource != channel_resources_.end())
+        // There is an existing channel in channel_resources_ created for reception with the remote locator as key. Use it.
+        channel = channel_resource->second;
+        // Add logical port to channel if it's not there yet
+        channel->add_logical_port(logical_port, rtcp_message_manager_.get());
+    }
+    // (Server-Client Topology - Client Side) OR LARGE DATA Topology with PDP discovery before TCP connection
+    else
+    {
+        // Get listening port (0 if client)
+        uint16_t listening_port = 0;
+        const TCPTransportDescriptor* config = configuration();
+        assert (config != nullptr);
+        if (!config->listening_ports.empty())
         {
-            channel = channel_resource->second;
+            listening_port = config->listening_ports.front();
         }
-        else
+
+        // If the remote physical port is higher than our listening port, a new CONNECT channel needs to be created and connected
+        // and the locator added to the send_resource_list.
+        // If the remote physical port is lower than our listening port, only the locator needs to be added to the send_resource_list.
+        if (IPLocator::getPhysicalPort(physical_locator) < listening_port)
```

Comment:
```suggestion
        if (IPLocator::getPhysicalPort(physical_locator) > listening_port)
```
I believe this OR the comments should be changed in order to be coherent.

Reply:
Totally right!